include/common/scaled-scene-buffer.h

```diff
@@ -17,6 +17,10 @@ struct scaled_scene_buffer_impl {
 		(struct scaled_scene_buffer *scaled_buffer, double scale);
 	/* Might be NULL or used for cleaning up */
 	void (*destroy)(struct scaled_scene_buffer *scaled_buffer);
+	/* Returns true if the two buffers are visually the same */
+	bool (*equal)(struct scaled_scene_buffer *scaled_buffer_a,
+		struct scaled_scene_buffer *scaled_buffer_b);
+	struct wl_list cached_buffers;
```

Comment:
We shouldn't put this list in the `impl` struct (which is in effect the method dispatch vtable).

Reply:
An alternative approach I could think of is using a static var per impl and then defining something like `scaled_font_buffer_init()` method that has to be called from somewhere like `server.c` and then does something like
```c
static wl_list cache;
[..]
void
scaled_font_buffer_init() {
	wl_list_init(&cache);
}
```

We could then supply that `&cache` to `scaled_scene_buffer_create()` whenever the impl is creating a new scaled buffer. It would cost one more pointer per `struct scaled_scene_buffer` but would be an alternative. Although I personally think that this is slightly more code compared to storing the list in the `impl` struct and doing a lazy `wl_list_init()`.

Reply:
I think a static variable at the start of the `scaled-scene-buffer.c` would be by far the clearest approach for me.

We could just init it in `scaled_scene_buffer_create()` with a quick check to make sure we only do it the first time.

Reply:
The cache belongs to the implementation of a scaled-scene-buffer, e.g. only scaled-font-buffer.c for now but each new implementation might want its own cache.

 > We could just init it in scaled_scene_buffer_create() with a quick check to make sure we only do it the first time.

That was the idea with the `if (!list.next) { wl_list_init(list); }`

Reply:
Yeah, it's just a fine line between "clear code" that someone like me (as reviewer) quickly gets - and using "tricks" to enable optimization.

Reply:
Yep. Maybe just `init_wl_list()` :smile:

Reply:
Or... how about just add a comment to describe this.

```
/*
 * As this is a shared list, it should only be initialized the first time.
 * When just `calloc()d` and not yet initialized, `prev`/`next` that is used to avoid
 * initizing a list in use
 */
```

Reply:
Fine with me as well. If we ever see the need to do this in other places we can always add a helper later.

Reply:
So, is adding a comment describing the intent of `impl.cached_buffer` enough for now? I'm not sure I understood the discussion correctly...

Another approach for one-shot initialization of `wl_list` would be defining a macro like this:

```c
#define WL_LIST_INIT(head) ((struct wl_list){.prev = (head), .next = (head)})

static struct scaled_scene_buffer_impl impl = {
	.cached_buffers = WL_LIST_INIT(&impl.cached_buffers),
};
```

I personally prefer this approach as it moves the task of initialization from runtime to compile-time and  eliminates additional checks.

Reply:
OK. I moved to the static var approach and added the following line in `scaled-font-buffer.c`:
```c
static struct wl_list cached_buffers = WL_LIST_INIT(&cached_buffers);
```